PebbleSharp.Core/AppMessage/AppMessagePacket.cs

```diff
@@ -0,0 +1,524 @@
+						}
+					}
+					else
+					{
+						throw new InvalidOperationException("Unknown tuple type");
+					}
+					index += l;
+					entry.Key = k;
+					Values.Add(entry);
+				}
+			}
+		}
+
+        public IList<IAppMessageDictionaryEntry> Values { get; set; }
+
+        public byte[] GetBytes()
+        {
+            if (Values != null && Values.Any())
+            {
+                var bytes = new List<byte>();
+                bytes.Add(Command);
+                bytes.Add(TransactionId);
+                bytes.AddRange(ApplicationId.Data);
+                bytes.Add((byte)Values.Count);
+                foreach (var tuple in Values)
+                {
+                    bytes.AddRange(tuple.PackedBytes);
+                }
+                return bytes.ToArray();
+            }
+            else
+            {
+                return new byte[0];
+            }
+        }
+    }
+
+    public interface IAppMessageDictionaryEntry
+    {
+        uint Key { get; set; }
+        PackedType PackedType { get; }
+        ushort Length { get; }
+        byte[] ValueBytes { get; set; }
+        byte[] PackedBytes { get; }
+    }
+
+    public enum Command:byte
+    {
+        Push=1
+    }
+
+    public enum PackedType:byte
+    {
+        Bytes=0,
+        String =1,
+        Unsigned =2,
+        Signed = 3
+    }
+
+    public abstract class AppMessageDictionaryEntry<T> : IAppMessageDictionaryEntry
+    {
+        public uint Key { get; set; }
+        public abstract PackedType PackedType { get; }
+        public abstract ushort Length { get; }
+
+        public virtual T Value { get; set; }
+        public abstract byte[] ValueBytes { get; set; }
+
+        public byte[] PackedBytes
+        {
+            get
+            {
+                var bytes = new List<byte>();
+                bytes.AddRange(BitConverter.GetBytes(Key));
+                bytes.Add((byte)PackedType);
+                bytes.AddRange(BitConverter.GetBytes(Length));
+                bytes.AddRange(ValueBytes);
+                return bytes.ToArray();
+            }
+        }
+    }
+
+    public class AppMessageUInt8 : AppMessageDictionaryEntry<byte>
+    {
+        public override PackedType PackedType
+        {
+            get { return PackedType.Unsigned; }
+        }
+
+        public override ushort Length
+        {
+            get { return sizeof(byte); }
+        }
+
+        public override byte[] ValueBytes
+        {
+            get { return new byte[] {Value}; }
+            set
+            {
+                if (value.Length == Length)
+                {
+                    Value = value[0];
+                }
+                else
+                {
+                    throw new InvalidOperationException("Incorrect # of bytes");
+                }
+            }
+        }
+
+        public override string ToString()
+        {
+            return Value.ToString();
+        }
```

Comment:
I would suggest implementing the ToString method in the base class an override it only in the classes where the implementation is not just calling Value.ToString(). I would also suggest using 
```C#
public override string ToString() => System.Convert.ToString(Value);
```
to avoid any potential null reference exceptions.